cpp/src/io/json/experimental/read_json.cpp

```diff
@@ -38,29 +39,47 @@ size_t sources_size(host_span<std::unique_ptr<datasource>> const sources,
   });
 }
 
-std::vector<uint8_t> ingest_raw_input(host_span<std::unique_ptr<datasource>> const& sources,
-                                      compression_type compression,
-                                      size_t range_offset,
-                                      size_t range_size)
+rmm::device_uvector<char> ingest_raw_input(host_span<std::unique_ptr<datasource>> const& sources,
+                                           compression_type compression,
+                                           size_t range_offset,
+                                           size_t range_size,
+                                           rmm::cuda_stream_view stream)
 {
   CUDF_FUNC_RANGE();
   // Iterate through the user defined sources and read the contents into the local buffer
   auto const total_source_size = sources_size(sources, range_offset, range_size);
-  auto buffer                  = std::vector<uint8_t>(total_source_size);
-
-  size_t bytes_read = 0;
-  for (const auto& source : sources) {
-    if (!source->is_empty()) {
-      auto data_size   = (range_size != 0) ? range_size : source->size();
-      auto destination = buffer.data() + bytes_read;
-      bytes_read += source->host_read(range_offset, data_size, destination);
-    }
-  }
 
   if (compression == compression_type::NONE) {
-    return buffer;
+    auto d_buffer     = rmm::device_uvector<char>(total_source_size, stream);
+    size_t bytes_read = 0;
+    for (const auto& source : sources) {
+      if (!source->is_empty()) {
+        auto data_size   = (range_size != 0) ? range_size : source->size();
+        auto destination = reinterpret_cast<uint8_t*>(d_buffer.data()) + bytes_read;
+        if (source->is_device_read_preferred(data_size)) {
+          bytes_read += source->device_read(range_offset, data_size, destination, stream);
+        } else {
+          auto const h_buffer = source->host_read(range_offset, data_size);
+          CUDF_CUDA_TRY(cudaMemcpyAsync(destination,
+                                        h_buffer->data(),
+                                        h_buffer->size(),
+                                        cudaMemcpyHostToDevice,
+                                        stream.value()));
+          stream.synchronize();
+          bytes_read += h_buffer->size();
+        }
+      }
+    }
+    return d_buffer;
+
   } else {
-    return decompress(compression, buffer);
+    auto buffer = std::vector<uint8_t>(total_source_size);
+    // only a single compressed source is supported
+    sources[0]->host_read(range_offset, total_source_size, buffer.data());
```

Comment:
Do we not want to do a device read here?

Reply:
Unfortunately not. Since the whole file is compressed as a single chunk, decompression is expected to be much faster on the CPU.
Once we add support for multiple compressed files, we can consider heuristics that choose where decompression is done based on the size and number of inputs. But I expect that this is a niche case.

Reply:
I’d consider adding that info (CPU is faster for single chunks) to a comment here. Otherwise I’d never realize that is the reasoning.

Reply:
added a comment